prov/sm2/src/sm2_progress.c

```diff
@@ -289,12 +435,98 @@ static int sm2_alloc_xfer_entry_ctx(struct sm2_ep *ep,
 static int sm2_progress_recv_msg(struct sm2_ep *ep,
 				 struct sm2_xfer_entry *xfer_entry)
 {
+	struct sm2_av *av =
+		container_of(ep->util_ep.av, struct sm2_av, util_av);
 	struct fid_peer_srx *peer_srx = sm2_get_peer_srx(ep);
 	struct fi_peer_rx_entry *rx_entry;
+	struct sm2_cma_data *cma_data;
 	struct sm2_av *sm2_av;
 	fi_addr_t addr;
+	uint64_t comp_flags;
 	int ret;
 
+	/* TODO - Switch on protocol before switching on op to avoid messy
+	 * checks like this */
+
+	/* We don't want to look for rx entries for xfer entries with
+	 * FI_SM2_CMA_HOST_TO_DEV and FI_SM2_CMA_HOST_TO_DEV_ACK flags*/
+	if (xfer_entry->hdr.proto == sm2_proto_cma) {
+		if (xfer_entry->hdr.proto_flags & FI_SM2_CMA_HOST_TO_DEV) {
+			/* Sender received IPC handle from receiver and needs to
+			 * do a device memcpy, so skip rx entry lookup */
```

Comment:
These are pretty much the same comment and I think it is self explanatory if you understand the protocol itself (it's fine to have a description of the protocol but I would put it in one place in the header instead of having 5-10 comments scattered throughout the code path explaining the flow of the code)